tests/test_construction_missing_annotations.py

```diff
@@ -0,0 +1,56 @@
+from __future__ import annotations
+
+from typing import ForwardRef, Optional
+
+import pytest
+
+from pydantic.main import BaseModel
+
+
+def test_missing_annotation_warning_defaults_to_true():
+
+    with pytest.raises(UserWarning, match='is not fully defined'):
+
+        Bar = ForwardRef('Bar')
+
+        class Foo(BaseModel):
+            a: Bar
+
+        class Bar(BaseModel):
+            foo: Optional[Foo] = None
+
+
+def test_missing_annotation_warning_disabled_in_config():
+
+    Bar = ForwardRef('Bar')
+
+    class Foo(BaseModel):
+        b: Bar
+
+        class Config:
+            warn_on_undefined_types = False
+
+    assert Foo.__pydantic_model_complete__ is False
+
+    class Bar(BaseModel):
+        foo: Optional[Foo] = None
+
+    assert Bar.__pydantic_model_complete__ is True
+
+    Foo.model_rebuild()
+    assert Foo.__pydantic_model_complete__ is True
+    assert Bar.__pydantic_model_complete__ is True
```

Comment:
This test should be revised to be a bit simpler. I gather that you are providing a reference in class `Foo` to `Bar`, the latter of which doesn't have an annotation. Thereafter, you provide `Bar` an annotation and expect to rebuild `Foo` without error. If yes, then building `Bar` to include `Foo` is unnecessarily complex. Why not provide `Bar` with simpler attributes, since this test just needs to validate that `Foo.__pydantic_model_complete__ is True` (Bar can be ignored...)

Reply:
Curious if you feel strongly 1 way or the other @samuelcolvin